python/cudf/dataframe/multiindex.py

```diff
@@ -208,29 +225,34 @@ def _get_row_major(self, df, row_tuple):
         out_index = DataFrame()
         # Select the last n-k columns where n is the number of source
         # levels and k is the length of the indexing tuple
-        for k in range(len(row_tuple), len(df.index.levels)):
+        size = 0
+        if not isinstance(row_tuple[0], (numbers.Number, slice)):
+            size = len(row_tuple)
+        for k in range(size, len(df.index.levels)):
             out_index.add_column(df.index.names[k],
                                  df.index.codes[df.index.codes.columns[k]])
         # If there's only one column remaining in the output index, convert
-        # it into a StringIndex and name the final index values according
+        # it into an Index and name the final index values according
         # to the proper codes.
         if len(out_index.columns) == 1:
             out_index = []
             for val in result.index.codes[result.index.codes.columns[len(result.index.codes.columns)-1]]:  # noqa: E501
                 out_index.append(result.index.levels[
                         len(result.index.codes.columns)-1][val])
-            # TODO: Warning! The final index column could be arbitrarily
-            # ordered integers, not Strings, so we need to check for that
-            # dtype and produce a GenericIndex instead of a StringIndex
-            out_index = StringIndex(out_index)
+            out_index = as_index(out_index)
             out_index.name = result.index.names[len(result.index.names)-1]
             result.index = out_index
         else:
-            # Otherwise pop the leftmost levels, names, and codes from the
-            # source index until it has the correct number of columns (n-k)
-            if(len(out_index.columns)) > 0:
+            if len(result) == 1 and size == 0 and slice_access is False:
+                # If the final result is one row and it was not mapped into
+                # directly
+                result = result.T
```

Comment:
I also worried about this line, but it'll only be a for a single row, guaranteed.